src/arithmetic/algebraic_expression/algebraic_expression_optimization.c

```diff
@@ -24,20 +25,20 @@ static void _AlgebraicExpression_CollectOperands(AlgebraicExpression *root,
 	switch(root->type) {
 	case AL_OPERAND:
 		*operands = array_append(*operands, AlgebraicExpression_Clone(root));
+		root->operand.should_free = false; // The caller is the new owner of this operand.
```

Comment:
Can you please explain this addition?

Reply:
This was necessary because transposition was being performed before the `FlattenMultiplications` optimization, which frees and replaces the original expression. This caused constructed matrices to be freed prematurely, and left invalid pointers in the new expression.

Since the transpose optimizations are now done in two parts, this line will not currently cause any changes, but I think logically it belongs!